backend/src/router.rs

```diff
@@ -0,0 +1,321 @@
+use crate::prelude::*;
+use axum::body::HttpBody;
+use axum::routing::on;
+use axum::{handler::Handler, routing::MethodFilter, Router};
+use hyper::{Body, Method};
+use std::convert::Infallible;
+use std::ops::Deref;
+use thiserror::Error;
+use utoipa::openapi::PathItemType;
+use utoipa::OpenApi;
+
+#[derive(Debug, Error, PartialEq, Eq, PartialOrd, Ord)]
+pub enum RouteError {
+    #[error("No route found in OpenAPI scheme")]
+    NoRoute,
+    #[error("No method found for specified route in OpenAPI scheme")]
+    NoMethod,
+    #[error("No `operation_id` found")]
+    NoOperation,
+    #[error("OpenAPI's `operation_id` doesn't match handler's name")]
+    NoMatch,
+    #[error("Unexpected Hyper method - was it `Method::CONNECT`?")]
+    UnexpectedMethod,
+    #[error("Error occured during handler processing")]
+    InvalidHandler,
+}
+
+pub struct NoApi;
+
+impl<'a> ApiVersion<'a> for NoApi {}
+
+pub struct ApiV1;
+
+pub trait ApiVersion<'a> {
+    #[must_use]
+    fn to_path() -> &'a str {
+        ""
+    }
+}
+
+impl<'a> ApiVersion<'a> for ApiV1 {
+    fn to_path() -> &'a str {
+        "/api/v1"
+    }
+}
+
+pub trait RouterApiExt<S = (), B = Body, E = Infallible> {
+    ///  Add API Route
+    ///   
+    fn api_route<'a, H, T, Version, ApiDocumentation>(
+        self,
+        path: &str,
+        method: Method,
+        handler: H,
+    ) -> Result<Self, RouteError>
+    where
+        H: Handler<T, S, B>,
+        T: 'static,
+        S: Send + Sync + 'static,
+        Version: ApiVersion<'a>,
+        ApiDocumentation: OpenApi,
+        Self: Sized;
+}
+
+impl<S, B> RouterApiExt<S, B, Infallible> for Router<S, B>
+where
+    B: HttpBody + Send + 'static,
+    S: Clone + Send + Sync + 'static,
+{
+    fn api_route<'a, H, T, Version, ApiDocumentation>(
+        self,
+        path: &str,
+        method: Method,
+        handler: H,
+    ) -> Result<Self, RouteError>
+    where
+        H: Handler<T, S, B>,
+        T: 'static,
+        S: Send + Sync + 'static,
+        Version: ApiVersion<'a>,
+        ApiDocumentation: OpenApi,
+        Self: Sized,
+    {
+        check_api::<_, _, _, H, Version, ApiDocumentation>(
+            path,
+            &*TryInto::<MethodWrapper<PathItemType>>::try_into(method.clone())?,
+        )?;
+        Ok(self.route(
+            path,
+            on(
+                *TryInto::<MethodWrapper<MethodFilter>>::try_into(method)?,
+                handler,
+            ),
+        ))
+    }
+}
+
+/// Check if the following route corresponds with `OpenAPI` declaration
+/// Relies on `operation_id` field, must NOT be changed on handler's declaration
+fn check_api<'a, T, S, B, H, Version, ApiDocumentation>(
+    path: &str,
+    method: &PathItemType,
+) -> Result<(), RouteError>
+where
+    H: Handler<T, S, B>,
+    T: 'static,
+    S: Send + Sync + 'static,
+    ApiDocumentation: OpenApi,
+    Version: ApiVersion<'a>,
+{
+    #[cfg(any(not(debug_assertions), not(feature = "swagger")))]
```

Comment:
This is  a bad place for this attribute. It is better to put `cfg` on the whole `check_api` function and on the place, where it is called